File: parser/src/test/java/com/github/fburato/highwheelmodules/bytecodeparser/classpath/ArchiveClassPathRootTest.java
package com.github.fburato.highwheelmodules.bytecodeparser.classpath;


import java.io.File;
import java.io.IOException;
import java.io.InputStream;
import java.util.Arrays;
import java.util.Collection;

import org.junit.jupiter.api.BeforeEach;

import com.github.fburato.highwheelmodules.model.bytecode.ElementName;
import org.junit.jupiter.api.Test;

import static org.assertj.core.api.Assertions.*;

public class ArchiveClassPathRootTest {

  private ArchiveClassPathRoot testee;

  @BeforeEach
  public void setup() throws Exception {
    // note mytests.jar is taken from
    // http://johanneslink.net/projects/cpsuite.jsp
    // assume GPL licence for this file. We do not link to any code within it
    // however
    this.testee = new ArchiveClassPathRoot(new File("mytests.jar"));
  }

  @Test
  public void classNamesShouldReturnAllClassNamesIArchive() {
    final Collection<ElementName> expected = Arrays.asList(
        ElementName.fromString("injar.p1.P1NoTest$InnerTest"),
        ElementName.fromString("injar.p1.P1NoTest"),
        ElementName.fromString("injar.p1.P1Test"),
        ElementName.fromString("injar.p2.P2Test"));
    assertThat(expected).isEqualTo(this.testee.classNames());
  }

  @Test
  public void getDataShouldReturnNullForUnknownClass() throws Exception {
    assertThat(this.testee.getData(ElementName.fromString("bar"))).isNull();
  }

  @Test
  public void getDataShouldReturnInputStreamForAKnownClass() throws Exception {
    assertThat(this.testee
        .getData(ElementName.fromString("injar.p1.P1Test"))).isNotNull();
  }

  @Test
  public void shouldReturnAReadableInputStream() {
    final byte b[] = new byte[100];
    try {
      final InputStream actual = this.testee.getData(ElementName
          .fromString("injar.p1.P1Test"));
      actual.read(b);
    } catch (final IOException ex) {
      fail("IO Exception should not be thrown");
    }
  }

}

File: parser/src/test/scala/com/github/fburato/highwheelmodules/bytecodeparser/classpath/ArchiveClassPathRootTest.scala
package com.github.fburato.highwheelmodules.bytecodeparser.classpath

import com.github.fburato.highwheelmodules.bytecodeparser.TryMatchers._
import com.github.fburato.highwheelmodules.model.bytecode.ElementName
import org.mockito.scalatest.MockitoSugar
import org.scalatest.OneInstancePerTest
import org.scalatest.matchers.should.Matchers
import org.scalatest.wordspec.AnyWordSpec

import java.io.{File, InputStream}
import java.nio.file.Paths
import scala.util.Success

class ArchiveClassPathRootTest
    extends AnyWordSpec
    with Matchers
    with MockitoSugar
    with OneInstancePerTest {

  private val testee = new ArchiveClassPathRoot(Paths.get("parser", "mytests.jar").toFile)

  "classNames" should {
    "return all classes in archive" in {
      testee.classNames should beSuccessWith[Seq[ElementName]] { classNames =>
        classNames should contain theSameElementsAs Seq(
          ElementName.fromString("injar.p1.P1NoTest$InnerTest"),
          ElementName.fromString("injar.p1.P1NoTest"),
          ElementName.fromString("injar.p1.P1Test"),
          ElementName.fromString("injar.p2.P2Test")
        )
      }
    }
  }

  "getData" should {
    "return empty for unknown class" in {
      testee.getData(ElementName.fromString("bar")) shouldEqual Success(None)
    }

    "return input stream for known class" in {
      testee.getData(ElementName.fromString("injar.p1.P1Test")) should beSuccessWith[Option[InputStream]](
        result => result should not be empty
      )
    }

    "return a readable input stream" in {
      val bytes = new Array[Byte](100)

      testee
        .getData(ElementName.fromString("injar.p1.P1Test"))
        .map(maybeStream => maybeStream.get.read(bytes)) should beSuccess
    }
  }
}
